deepmd/dpmodel/model/transform_output.py

```diff
@@ -116,6 +118,96 @@ def communicate_extended_output(
                 else:
                     # name holders
                     new_ret[kk_derv_r] = None
+                if vdef.r_hessian:
+                    kk_hess = get_hessian_name(kk)
+                    if model_ret[kk_hess] is not None:
+                        # jax only
+                        if array_api_compat.is_jax_array(force):
```

Comment:
would it be better to implement the hessian scatter sum to somewhere in the jax module and import it here. 
otherwise the code would become difficult to read when more array api compatible backends are added. 

The same comment for the force part.

Reply:
I think I could try to convert it to a common one.

Reply:
> I think I could try to convert it to a common one.

#4654 tries to convert scatter_sum to a common one, but it needs Python 3.10.

Reply:
I've done an independent refactor in this PR, which doesn't conflict with #4654.